// ng-dynamic-component/esm2020/lib/dynamic-io/dynamic-io.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentOutletInjectorModule } from '../component-injector/component-outlet-injector.module';
import { DynamicIoDirective } from './dynamic-io.directive';
import * as i0 from "@angular/core";
export class DynamicIoModule {
}
/** @nocollapse */ /** @nocollapse */ DynamicIoModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicIoModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ /** @nocollapse */ DynamicIoModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicIoModule, declarations: [DynamicIoDirective], imports: [CommonModule], exports: [DynamicIoDirective, ComponentOutletInjectorModule] });
/** @nocollapse */ /** @nocollapse */ DynamicIoModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicIoModule, imports: [[CommonModule], ComponentOutletInjectorModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicIoModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [DynamicIoDirective, ComponentOutletInjectorModule],
                    declarations: [DynamicIoDirective],
                }]
        }] });
