// ng-dynamic-component/esm2020/lib/dynamic.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DynamicIoModule } from './dynamic-io/dynamic-io.module';
import { DynamicComponent } from './dynamic.component';
import * as i0 from "@angular/core";
export class DynamicModule {
}
/** @nocollapse */ /** @nocollapse */ DynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ /** @nocollapse */ DynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicModule, declarations: [DynamicComponent], imports: [CommonModule, DynamicIoModule], exports: [DynamicComponent, DynamicIoModule] });
/** @nocollapse */ /** @nocollapse */ DynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicModule, imports: [[CommonModule, DynamicIoModule], DynamicIoModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DynamicIoModule],
                    exports: [DynamicComponent, DynamicIoModule],
                    declarations: [DynamicComponent],
                }]
        }] });
