// ng-dynamic-component/esm2020/lib/dynamic-attributes/dynamic-attributes.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentOutletInjectorModule } from '../component-injector/component-outlet-injector.module';
import { DynamicAttributesDirective } from './dynamic-attributes.directive';
import * as i0 from "@angular/core";
export class DynamicAttributesModule {
}
/** @nocollapse */ /** @nocollapse */ DynamicAttributesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicAttributesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ /** @nocollapse */ DynamicAttributesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicAttributesModule, declarations: [DynamicAttributesDirective], imports: [CommonModule], exports: [DynamicAttributesDirective, ComponentOutletInjectorModule] });
/** @nocollapse */ /** @nocollapse */ DynamicAttributesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicAttributesModule, imports: [[CommonModule], ComponentOutletInjectorModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicAttributesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [DynamicAttributesDirective, ComponentOutletInjectorModule],
                    declarations: [DynamicAttributesDirective],
                }]
        }] });
