// ng-dynamic-component/esm2020/lib/dynamic-directives/dynamic-directives.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentOutletInjectorModule } from '../component-injector/component-outlet-injector.module';
import { DynamicDirectivesDirective } from './dynamic-directives.directive';
import * as i0 from "@angular/core";
export class DynamicDirectivesModule {
}
/** @nocollapse */ /** @nocollapse */ DynamicDirectivesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicDirectivesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ /** @nocollapse */ DynamicDirectivesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicDirectivesModule, declarations: [DynamicDirectivesDirective], imports: [CommonModule], exports: [DynamicDirectivesDirective, ComponentOutletInjectorModule] });
/** @nocollapse */ /** @nocollapse */ DynamicDirectivesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicDirectivesModule, imports: [[CommonModule], ComponentOutletInjectorModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DynamicDirectivesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [DynamicDirectivesDirective, ComponentOutletInjectorModule],
                    declarations: [DynamicDirectivesDirective],
                }]
        }] });
